pkg/sensors/accelerator/gpu/source/gpu_dcgm.go

```diff
@@ -0,0 +1,372 @@
+		// necessary to calculate the energy consumption for the entire waiting period
+		energy := uint32(uint64(power) * config.SamplePeriodSec)
+		gpuEnergy = append(gpuEnergy, energy)
+	}
+	return gpuEnergy
+}
+
+func (d *GPUDcgm) GetGpus() map[string]interface{} {
+	return d.devices
+}
+
+func (d *GPUDcgm) GetProcessResourceUtilizationPerDevice(device interface{}, deviceName string, since time.Duration) (map[uint32]ProcessUtilizationSample, error) {
+	processAcceleratorMetrics := map[uint32]ProcessUtilizationSample{}
+
+	if device == nil { // this is a MIG device, it is already tracked in the parent device
+		return processAcceleratorMetrics, nil
+	}
+
+	var vals, miVals []dcgm.FieldValue_v1
+	var err error
+
+	klog.V(debugLevel).Infof("Device %v\n", deviceName)
+
+	deviceIndex, strErr := strconv.Atoi(deviceName)
+	if strErr != nil {
+		klog.V(debugLevel).Infof("failed to convert %q to an integer: %v", deviceName, strErr)
+		return processAcceleratorMetrics, strErr
+	}
+	vals, err = dcgm.GetLatestValuesForFields(uint(deviceIndex), deviceFields)
+	if err != nil {
+		klog.V(debugLevel).Infof("failed to get latest values for fields: %v", err)
+		return processAcceleratorMetrics, err
+	}
+	gpuSMActive := uint32(0)
+	if err == nil {
+		for i, val := range vals {
+			value := ToString(val)
+			label := deviceFieldsString[i]
+			if val.FieldId == ratioFields {
+				smUtil, _ := strconv.ParseFloat(value, 32)
+				gpuSMActive = uint32(smUtil * 100)
+			}
+			klog.V(debugLevel).Infof("Device %v Label %v Val: %v", deviceName, label, ToString(val))
+		}
+		klog.V(debugLevel).Infof("\n")
+	}
+	processInfo, ret := device.(nvml.Device).GetComputeRunningProcesses()
+	if ret != nvml.SUCCESS {
+		klog.V(debugLevel).Infof("failed to get running processes: %v", nvml.ErrorString(ret))
+		return processAcceleratorMetrics, fmt.Errorf("failed to get running processes: %v", nvml.ErrorString(ret))
+	}
+	for _, p := range processInfo {
+		// klog.V(debugLevel).Infof("pid: %d, memUtil: %d gpu instance id %d compute id %d\n", p.Pid, p.UsedGpuMemory, p.GpuInstanceId, p.ComputeInstanceId)
+		if p.GpuInstanceId > 0 { // this is a MIG, get it entity id and reads the related fields
+			entityName := gpuMigArray[deviceIndex][p.GpuInstanceId].EntityName
+			multiprocessorCountRatio := gpuMigArray[deviceIndex][p.GpuInstanceId].MultiprocessorCountRatio
+			mi := d.entities[entityName]
+			miVals, err = dcgm.EntityGetLatestValues(mi.EntityGroupId, mi.EntityId, deviceFields)
+			if err == nil {
+				for i, val := range miVals {
+					label := deviceFieldsString[i]
+					value := ToString(val)
+					klog.V(debugLevel).Infof("Device %v Label %v Val: %v", entityName, label, value)
+					if val.FieldId == ratioFields {
+						floatVal, _ := strconv.ParseFloat(value, 32)
+						// ratio of active multiprocessors to total multiprocessors
+						smUtil := uint32(floatVal * 100 * multiprocessorCountRatio)
```

Comment:
This variable represents tensor core utilization, not SM utilization.

It might be better to just use "util" since the variable will hold the utilization of different components in the future.